program/rust/src/rust_oracle.rs

```diff
@@ -803,3 +836,77 @@ pub fn del_product(
 
     Ok(())
 }
+
+
+pub fn upd_permissions(
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    instruction_data: &[u8],
+) -> ProgramResult {
+    let [funding_account, program_account, programdata_account, permissions_account, system_program] =
+        match accounts {
+            [v, w, x, y, z] => Ok([v, w, x, y, z]),
+            _ => Err(ProgramError::InvalidArgument),
+        }?;
+
+    let cmd_args = load::<UpdPermissionsArgs>(instruction_data)?;
+
+    check_valid_funding_account(funding_account)?;
+    check_is_valid_upgradeauthority_program_programdata_triplet(
+        funding_account,
+        program_account,
+        programdata_account,
+        program_id,
+    )?;
+    let bump = check_pda(
+        permissions_account,
+        &[PERMISSIONS_SEED.as_bytes()],
+        program_id,
+    )?;
+    pyth_assert(
+        check_id(system_program.key),
+        OracleError::InvalidSystemAccount.into(),
+    )?;
+
+
+    // Create account if it doesn't exist
+    if permissions_account.lamports() == 0 {
+        create_pda_account(
+            funding_account,
+            permissions_account,
+            system_program,
+            PermissionAccount::minimum_size(),
+            program_id,
+            &[PERMISSIONS_SEED.as_bytes(), &[bump]],
+        )?;
+        check_valid_fresh_account(permissions_account)?;
+        initialize_pyth_account_checked::<PermissionAccount>(
+            permissions_account,
+            cmd_args.header.version,
+        )?;
+    }
```

Comment:
This can be factored out into a helper function.

Reply:
Done